File: buffer-overflow-demo/lower.rs
use std::{env::args_os, process::exit};

fn main() {
    match args_os().nth(1) {
        None => {
            eprint!("usage: <program> <input>");
            exit(1);
        }
        Some(arg) => {
            let mut input: Vec<u8> = arg.into_encoded_bytes();
            lowercase_print(&mut input);
        }
    }
}

fn lowercase_print(input_str: &mut [u8]) {
    let mut unmodified = [0; 32];
    strcpy(&mut unmodified, input_str);
    str_to_lowercase(input_str);
    println!(
        "The string '{}' in lowercase is: '{}'",
        slice2str(&unmodified),
        slice2str(input_str)
    );
}

// NOTE: Rust has no equivalent strcpy()
fn strcpy(dest: &mut [u8], src: &[u8]) {
    for (i, character) in src.iter().enumerate() {
        dest[i] = *character;
    }
}

fn str_to_lowercase(string: &mut [u8]) {
    let mut temp = vec![0; string.len()];
    temp.clone_from_slice(string);
    for (i, character) in temp.iter().enumerate() {
        string[i] = character.to_ascii_lowercase();
    }
}

// ---------- Helper -------------
fn slice2str(slice: &[u8]) -> String {
    slice.iter().map(|code| *code as char).collect()
}

File: buffer-overflow-demo/lower.c
#include <ctype.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void str_to_lowercase(char *str) {
    int length = strlen(str);
    for (int i = 0; i < length; ++i)
        str[i] = tolower(str[i]);
}

void lowercase_print(char *input_str) {
    char unmodified[32];
    strcpy(unmodified, input_str);
    str_to_lowercase(input_str);
    printf("The string '%s' in lowercase is: '%s'\n", unmodified, input_str);
}

void secret_function() {
    printf("\n!!!You have reached the secret code!!!\n");
    system("/bin/sh");
}

int main(int argc, char *argv[]) {
    if (argc != 2) {
        perror("usage: <program> <input>\n");
        return 1;
    }
    lowercase_print(argv[1]);
}
